controllers/service_controller.go

```diff
@@ -103,8 +108,47 @@ func (r *serviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 	if !svc.DeletionTimestamp.IsZero() {
+		tgName := latticestore.TargetGroupName(svc.Name, svc.Namespace)
+		tgs := r.datastore.GetTargetGroupsByName(tgName)
+		for _, tg := range tgs {
+			r.log.Debugf("deletion request for tgName: %s: at timestamp: %s", tg.TargetGroupKey.Name, svc.DeletionTimestamp)
+			r.reconcileTargetsResource(ctx, svc, tg.TargetGroupKey.RouteName)
+		}
 		r.finalizerManager.RemoveFinalizers(ctx, svc, serviceFinalizer)
+	} else {
+		if err := r.finalizerManager.AddFinalizers(ctx, svc, serviceFinalizer); err != nil {
+			r.eventRecorder.Event(svc, corev1.EventTypeWarning, k8s.ServiceEventReasonFailedAddFinalizer, fmt.Sprintf("failed and finalizer: %s", err))
+		}
 	}
+
 	r.log.Infow("reconciled", "name", req.Name)
 	return ctrl.Result{}, nil
 }
+
+func (r *serviceReconciler) reconcileTargetsResource(ctx context.Context, svc *corev1.Service, routename string) {
+	if err := r.finalizerManager.AddFinalizers(ctx, svc, serviceFinalizer); err != nil {
```

Comment:
I don't think this part of adding finalizer is needed right now

Reply:
Old `reconcileTargetsResource() ` has the `AddFinalizers(ctx, svc, serviceFinalizer)`, I can remove if its only use to handle delete k8sService